// ngx-scrollbar/esm2022/lib/scrollbar/scrollbar-adapter.mjs
import { Directive, ViewChild, effect, inject, NgZone } from '@angular/core';
import { merge } from 'rxjs';
import { NG_SCROLLBAR } from '../utils/scrollbar-base';
import * as i0 from "@angular/core";
// @dynamic
export class ScrollbarAdapter {
    constructor() {
        // Zone reference
        this.zone = inject(NgZone);
        // Host component reference
        this.cmp = inject(NG_SCROLLBAR);
        effect((onCleanup) => {
            if (this.cmp.interactionDisabled()) {
                this.pointerEventsSub?.unsubscribe();
            }
            else {
                this.zone.runOutsideAngular(() => {
                    this.pointerEventsSub = merge(
                    // Activate scrollbar thumb drag event
                    this.thumb.dragged, 
                    // Activate scrollbar track click event
                    this.track.dragged).subscribe();
                });
            }
            onCleanup(() => this.pointerEventsSub?.unsubscribe());
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ScrollbarAdapter, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.1", type: ScrollbarAdapter, viewQueries: [{ propertyName: "sticky", first: true, predicate: ["sticky"], descendants: true, static: true }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ScrollbarAdapter, decorators: [{
            type: Directive
        }], ctorParameters: () => [], propDecorators: { sticky: [{
                type: ViewChild,
                args: ['sticky', { static: true }]
            }] } });
