// ngx-scrollbar/esm2022/lib/viewport/viewport-adapter.mjs
import { ViewportClasses } from '../ng-scrollbar.model';
export class ViewportAdapter {
    /** Viewport clientHeight */
    get offsetHeight() {
        return this.nativeElement.offsetHeight;
    }
    /** Viewport clientWidth */
    get offsetWidth() {
        return this.nativeElement.offsetWidth;
    }
    /** Viewport scrollTop */
    get scrollTop() {
        return this.nativeElement.scrollTop;
    }
    /** Viewport scrollLeft */
    get scrollLeft() {
        return this.nativeElement.scrollLeft;
    }
    /** Content height, falls back to scroll height */
    get contentHeight() {
        return this.contentWrapperElement?.offsetHeight;
    }
    /** Content width, falls back to scroll height */
    get contentWidth() {
        return this.contentWrapperElement?.offsetWidth;
    }
    /** The horizontal remaining scrollable distance */
    get scrollMaxX() {
        return this.contentWidth - this.offsetWidth;
    }
    /** The vertical remaining scrollable distance */
    get scrollMaxY() {
        return this.contentHeight - this.offsetHeight;
    }
    constructor(nativeElement) {
        this.nativeElement = nativeElement;
        nativeElement.classList.add(ViewportClasses.Viewport);
    }
    /**
     * Initialize viewport
     */
    init(contentSelector, spacerSelector) {
        // When integrating the scrollbar with virtual scroll, the content wrapper will have fake size,
        // and a spacer element will have the real size
        // Therefore, if spaceElement is provided, it will be observed instead of the content wrapper
        if (spacerSelector) {
            // Set relative position on the spacer element to enable the functionality of sticky for the scrollbars
            spacerSelector.style.position = 'relative';
            this.contentWrapperElement = spacerSelector;
        }
        let realContentWrapper = contentSelector ?? this.nativeElement?.firstElementChild;
        // Add content wrapper class
        realContentWrapper?.classList.add(ViewportClasses.Content);
        // If spacer is not provided, set it as the content wrapper
        if (!this.contentWrapperElement && realContentWrapper) {
            this.contentWrapperElement = realContentWrapper;
        }
    }
    /**
     * Scroll viewport vertically
     */
    scrollYTo(value) {
        this.nativeElement.scrollTop = value;
    }
    /**
     * Scroll viewport horizontally
     */
    scrollXTo(value) {
        this.nativeElement.scrollLeft = value;
    }
}
