// ngx-scrollbar/esm2022/lib/thumb/thumb.mjs
import { Directive, effect } from '@angular/core';
import { RtlScrollAxisType } from '@angular/cdk/platform';
import { ThumbAdapter } from './thumb-adapter';
import * as i0 from "@angular/core";
export class ThumbXDirective extends ThumbAdapter {
    get clientProperty() {
        return 'clientX';
    }
    get pageProperty() {
        return 'pageX';
    }
    get viewportScrollMax() {
        return this.cmp.viewport.scrollMaxX;
    }
    get dragStartOffset() {
        return this.clientRect.left + this.document.defaultView.pageXOffset;
    }
    get offset() {
        return this.clientRect.left;
    }
    get size() {
        return this.nativeElement.clientWidth;
    }
    constructor() {
        super();
        this.axis = 'x';
        effect(() => {
            if (this.cmp.direction() === 'rtl') {
                if (this.cmp.rtlScrollAxisType === RtlScrollAxisType.NEGATED) {
                    this.handleDrag = (position, scrollMax) => -(scrollMax - position);
                }
                if (this.cmp.rtlScrollAxisType === RtlScrollAxisType.INVERTED) {
                    this.handleDrag = (position, scrollMax) => position - scrollMax;
                }
            }
            else {
                this.handleDrag = (position) => position;
            }
        });
    }
    scrollTo(position) {
        this.cmp.viewport.scrollXTo(position);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ThumbXDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.1", type: ThumbXDirective, isStandalone: true, selector: "[scrollbarThumbX]", providers: [{ provide: ThumbAdapter, useExisting: ThumbXDirective }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ThumbXDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[scrollbarThumbX]',
                    providers: [{ provide: ThumbAdapter, useExisting: ThumbXDirective }]
                }]
        }], ctorParameters: () => [] });
export class ThumbYDirective extends ThumbAdapter {
    constructor() {
        super(...arguments);
        this.axis = 'y';
        this.handleDrag = (position) => position;
    }
    get pageProperty() {
        return 'pageY';
    }
    get viewportScrollMax() {
        return this.cmp.viewport.scrollMaxY;
    }
    get clientProperty() {
        return 'clientY';
    }
    get dragStartOffset() {
        return this.clientRect.top + this.document.defaultView.pageYOffset;
    }
    get offset() {
        return this.clientRect.top;
    }
    get size() {
        return this.nativeElement.clientHeight;
    }
    scrollTo(position) {
        this.cmp.viewport.scrollYTo(position);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ThumbYDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.1", type: ThumbYDirective, isStandalone: true, selector: "[scrollbarThumbY]", providers: [{ provide: ThumbAdapter, useExisting: ThumbYDirective }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ThumbYDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[scrollbarThumbY]',
                    providers: [{ provide: ThumbAdapter, useExisting: ThumbYDirective }]
                }]
        }] });
