// ngx-scrollbar/esm2022/lib/utils/scrollbar-manager.mjs
import { Injectable, Inject, inject, Optional, signal, PLATFORM_ID } from '@angular/core';
import { DOCUMENT, isPlatformBrowser } from '@angular/common';
import { getRtlScrollAxisType } from '@angular/cdk/platform';
import { NG_SCROLLBAR_OPTIONS } from '../ng-scrollbar.model';
import * as i0 from "@angular/core";
const defaultOptions = {
    scrollTimelinePolyfill: 'https://flackr.github.io/scroll-timeline/dist/scroll-timeline.js',
    trackClass: '',
    thumbClass: '',
    orientation: 'auto',
    appearance: 'standard',
    visibility: 'native',
    position: 'native',
    clickScrollDuration: 50,
    sensorThrottleTime: 0,
    disableSensor: false,
    disableInteraction: false
};
export class ScrollbarManager {
    constructor(options) {
        this.isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
        this.document = inject(DOCUMENT);
        this.window = this.document.defaultView;
        this.globalOptions = {};
        this.rtlScrollAxisType = getRtlScrollAxisType();
        this.scrollTimelinePolyfill = signal(null);
        this.globalOptions = options ? { ...defaultOptions, ...options } : defaultOptions;
        if (this.isBrowser && !this.window['ScrollTimeline'] && !CSS.supports('animation-timeline', 'scroll()')) {
            this.initPolyfill();
        }
    }
    async initPolyfill() {
        try {
            // Create a script element
            const script = this.document.createElement('script');
            script.src = this.globalOptions.scrollTimelinePolyfill;
            // Wait for the script to load
            await new Promise((resolve, reject) => {
                script.onload = resolve;
                script.onerror = reject;
                this.document.head.appendChild(script);
            });
            // Once loaded, access and execute the function attached to the window object
            if (this.window['ScrollTimeline']) {
                this.scrollTimelinePolyfill.set(window['ScrollTimeline']);
            }
            else {
                console.error('ScrollTimeline is not attached to the window object.');
            }
        }
        catch (error) {
            console.error('Error loading ScrollTimeline script:', error);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ScrollbarManager, deps: [{ token: NG_SCROLLBAR_OPTIONS, optional: true }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ScrollbarManager, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: ScrollbarManager, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [NG_SCROLLBAR_OPTIONS]
                }] }] });
