// ngx-scrollbar/fesm2022/ngx-scrollbar-utils.mjs
import * as i0 from '@angular/core';
import { inject, effect, Directive } from '@angular/core';
import { MatSelect } from '@angular/material/select';
import { NG_SCROLLBAR, NgScrollbarExt } from 'ngx-scrollbar';

class MatSelectViewport {
    constructor() {
        this.scrollbar = inject(NG_SCROLLBAR);
        this.matSelect = inject(MatSelect);
        effect(() => {
            if (this.scrollbar.isVerticallyScrollable() && this.matSelect.panelOpen) {
                const selected = this.matSelect.selected;
                if (selected) {
                    const element = Array.isArray(selected) ? selected[0]._getHostElement() : selected._getHostElement();
                    const height = this.scrollbar.nativeElement.clientHeight;
                    this.scrollbar.scrollTo({
                        top: element.offsetTop + element.offsetHeight - height
                    });
                }
            }
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: MatSelectViewport, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.1", type: MatSelectViewport, isStandalone: true, selector: "ng-scrollbar[matSelectViewport], ng-scrollbar[mat-select-viewport]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: MatSelectViewport, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-scrollbar[matSelectViewport], ng-scrollbar[mat-select-viewport]'
                }]
        }], ctorParameters: () => [] });

class NgScrollbarCdkVirtualScroll {
    constructor() {
        this.scrollbar = inject(NgScrollbarExt);
        this.scrollbar.externalViewport = '.cdk-virtual-scroll-viewport';
        this.scrollbar.externalSpacer = '.cdk-virtual-scroll-spacer';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: NgScrollbarCdkVirtualScroll, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.1", type: NgScrollbarCdkVirtualScroll, isStandalone: true, selector: "ng-scrollbar[cdkVirtualScrollViewport]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: NgScrollbarCdkVirtualScroll, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-scrollbar[cdkVirtualScrollViewport]'
                }]
        }], ctorParameters: () => [] });

/**
 * Generated bundle index. Do not edit.
 */

export { MatSelectViewport, NgScrollbarCdkVirtualScroll };
