// ngx-scrollbar/esm2022/smooth-scroll/smooth-scroll.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./smooth-scroll-manager";
export class SmoothScroll {
    constructor(element, smoothScroll) {
        this.element = element;
        this.smoothScroll = smoothScroll;
    }
    scrollTo(options) {
        return this.smoothScroll.scrollTo(this.element, options);
    }
    scrollToElement(target, options) {
        return this.smoothScroll.scrollToElement(this.element, target, options);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: SmoothScroll, deps: [{ token: i0.ElementRef }, { token: i1.SmoothScrollManager }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.2.1", type: SmoothScroll, isStandalone: true, selector: "[smoothScroll], [smooth-scroll]", exportAs: ["smoothScroll"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.2.1", ngImport: i0, type: SmoothScroll, decorators: [{
            type: Directive,
            args: [{
                    selector: '[smoothScroll], [smooth-scroll]',
                    exportAs: 'smoothScroll',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i1.SmoothScrollManager }] });
